### ISSUE TITLE
Memory Leak & Erlang Crash 

### ISSUE THREAD

**Issue by jnmayur** [processone/ejabberd#3521]
## Environment

- ejabberd version: 20.01
- Erlang version: 11.0.3
- OS: 16.04.7 LTS (Xenial Xerus)
- Installed from: source 

## Configuration (only if needed): grep -Ev '^$|^\s*#' ejabberd.yml

```##
###              ejabberd configuration file
###
### The parameters used in this configuration file are explained at
###
###       https://docs.ejabberd.im/admin/configuration
###
### The configuration file is written in YAML.
### *******************************************************
### *******           !!! WARNING !!!               *******
### *******     YAML IS INDENTATION SENSITIVE       *******
### ******* MAKE SURE YOU INDENT SECTIONS CORRECTLY *******
### *******************************************************
### Refer to http://en.wikipedia.org/wiki/YAML for the brief description.
###
hosts:
  - example
loglevel: info
## If you already have certificates, list them here
certfiles:
     - /etc/ssl/ejabberd/***.cert
     - /etc/ssl/ejabberd/***.pem
    
#certfiles:
#   	- /etc/ssl/ejabberd/***.pem
#  - /etc/letsencrypt/live/domain.tld/privkey.pem
listen:
  -
    port: 5222
    ip: "::"
    module: ejabberd_c2s
    max_stanza_size: 262144
    shaper: c2s_shaper
    access: c2s
    starttls_required: true
  -
    port: 5269
    ip: "::"
    module: ejabberd_s2s_in
    max_stanza_size: 524288
  -
    port: 8443
    ip: "::"
    module: ejabberd_http
    tls: true
    #certfile: "/etc/ssl/ejabberd/***.cert"
    request_handlers:
      /websockets: ejabberd_http_ws
      /bosh: mod_bosh
      /admin: ejabberd_web_admin
  -
    port: 5281
    ip: "::"
    module: ejabberd_http
    request_handlers:
      /api: mod_http_api
  -
    port: 5280
    ip: "::"
    module: ejabberd_http
    request_handlers:
      /bosh: mod_bosh
      /websockets: ejabberd_http_ws
      /admin: ejabberd_web_admin
     # /.well-known/acme-challenge: ejabberd_acme
  -
    port: 3478
    transport: udp
    module: ejabberd_stun
    use_turn: true
    ## The server's public IPv4 address:
    # turn_ip: 203.0.113.3
  -
    port: 1883
    ip: "::"
    module: mod_mqtt
    backlog: 1000
s2s_use_starttls: optional
acl:
  admin:
    user: "admin@**.***.com"
    ip: "0.0.0.0"
  local:
    user_regexp: ""
  loopback:
    ip:
      - "127.0.0.1/8"
      - "::1"
access:
  configure:
    admin: allow
access_rules:
  local:
    allow: local
  c2s:
    deny: blocked
    allow: all
  announce:
    allow: admin
  configure:
    allow: admin
  muc_create:
    allow: loopback
  pubsub_createnode:
    allow: local
  trusted_network:
    allow: loopback
api_permissions:
  "console commands":
    from:
      - ejabberd_ctl
    who: all
    what: "*"
  "admin access":
    who:
      - access:
          - allow:
            - acl: admin
      - oauth:
        - scope: "ejabberd:admin"
        - access:
          - allow:
              - acl: admin
    what:
      - "*"
      - "!stop"
      - "!start"
  "public commands":
    who: all
    what:
      - "status"
      - "register"
      - "connected_users_number"
      - "create_room"
      - "set_room_affiliation"
      - "subscribe_room"
      - "unregister"
shaper:
  normal: 1000
  fast: 50000
shaper_rules:
  max_user_sessions: 10
  max_user_offline_messages:
    5000: admin
    100: all
  c2s_shaper:
    none: admin
    normal: all
  s2s_shaper: fast
modules:
  mod_adhoc: {}
  mod_admin_extra: {}
  mod_announce:
    access: announce
  mod_avatar: {}
  mod_blocking: {}
  mod_bosh: {}
  mod_caps: {}
  mod_carboncopy: {}
  mod_client_state: {}
  mod_configure: {}
  mod_disco: {}
  mod_fail2ban: {}
  mod_http_api: {}
  mod_http_upload:
    put_url: https://@HOST@:5443/upload
  mod_last: {}
  mod_mam:
    ## Mnesia is limited to 2GB, better to use an SQL backend
    ## For small servers SQLite is a good fit and is very easy
    ## to configure. Uncomment this when you have SQL configured:
    db_type: sql
    assume_mam_usage: true
    default: always
  mod_mqtt: {}
  mod_muc:
    access:
      - allow
    access_admin:
      - allow: admin
    access_create: muc_create
    access_persistent: muc_create
    access_mam:
      - allow
    default_room_options:
      mam: true
      persistent: true
      allow_subscription: true
    db_type: sql
    preload_rooms: false
  mod_muc_admin: {}
  mod_offline:
    store_empty_body: unless_chat_state
    store_groupchat: true
    access_max_user_messages: max_user_offline_messages
  mod_ping:
    send_pings: true
    ping_interval: 1 min
    timeout_action: kill
  mod_privacy: {}
  mod_private: {}
  mod_proxy65:
    access: local
    max_connections: 5
  mod_pubsub:
    access_createnode: pubsub_createnode
    plugins:
      - flat
      - pep
    force_node_config:
      ## Avoid buggy clients to make their bookmarks public
      storage:bookmarks:
        access_model: whitelist
  mod_push: {}
  mod_push_keepalive: {}
  mod_register:
    ## Only accept registration requests from the "trusted"
    ## network (see access_rules section above).
    ## Think twice before enabling registration from any
    ## address. See the Jabber SPAM Manifesto for details:
    ## https://github.com/ge0rg/jabber-spam-fighting-manifesto
    ip_access: trusted_network
  mod_roster:
    versioning: true
    db_type: sql
  mod_s2s_dialback: {}
  mod_shared_roster: {}
  mod_stream_mgmt:
    resume_timeout: 1 min
    resend_on_timeout: if_offline
  mod_stun_disco: {}
  mod_vcard: {}
  mod_vcard_xupdate: {}
  mod_version:
    show_os: false
  mod_offline_http_post:
    auth_token: "secret"
    post_url: "***"
    confidential: true
  mod_read_markers:
    auth_token: "secret"
    post_url: "***"
    confidential: true
sql_type: mysql
sql_server: "***"
sql_database: "ejabberd_db"
sql_username: "***"
sql_password: "***"
sql_port: 3306
default_db: sql
auth_method: sql
### Local Variables:
### mode: yaml
### End:

...
```

## Errors from error.log/crash.log

 pid: <0.314.0>
    registered_name: ext_mod
    exception exit: {aborted,{node_not_running,ejabberd@localhost}}
      in function  mnesia:abort/1 (mnesia.erl, line 361)
      in call from mnesia_tm:prepare_items/5 (mnesia_tm.erl, line 1225)
      in call from mnesia_tm:dirty/2 (mnesia_tm.erl, line 1067)
      in call from lists:foreach/2 (lists.erl, line 1338)
      in call from ejabberd_commands:unregister_commands/1 (src/ejabberd_commands.erl, line 145)
      in call from gen_server:try_terminate/3 (gen_server.erl, line 718)
      in call from gen_server:terminate/10 (gen_server.erl, line 903)
    ancestors: [ejabberd_sup,<0.116.0>]
    message_queue_len: 0
    messages: []
    links: []
    dictionary: []
    trap_exit: true
    status: running
    heap_size: 4185
    stack_size: 28
    reductions: 55143
  neighbours:
2021-02-07 05:38:18.259107+00:00 [error] <0.245.0>@supervisor:do_terminate/2:844 SUPERVISOR REPORT:
    supervisor: {local,ejabberd_sup}
    errorContext: shutdown_error
    reason: {aborted,{node_not_running,ejabberd@localhost}}
    offender: [{pid,<0.314.0>},
               {id,ext_mod},
               {mfargs,{ext_mod,start_link,[]}},
               {restart_type,permanent},
               {shutdown,60000},
               {child_type,worker}]
2021-02-07 05:38:18.259788+00:00 [error] <0.302.0>@ejabberd_router_mnesia:transaction/1:222 Mnesia transaction failed: {node_not_running,ejabberd@localhost}
2021-02-07 05:38:18.259996+00:00 [error] <0.302.0>@ejabberd_router:unregister_route/2:209 Failed to unregister route pigeon.urbancompany.com: db_failure
2021-02-07 05:38:18.260378+00:00 [error] <0.286.0>@ejabberd_router_mnesia:transaction/1:222 Mnesia transaction failed: {node_not_running,ejabberd@localhost}
2021-02-07 05:38:18.260636+00:00 [error] <0.284.0>@ejabberd_router_mnesia:transaction/1:222 Mnesia transaction failed: {node_not_running,ejabberd@localhost}
2021-02-07 05:38:18.260812+00:00 [error] <0.284.0>@ejabberd_router:unregister_route/2:209 Failed to unregister route pigeon.urbancompany.com: db_failure
2021-02-07 05:38:19.083683+00:00 [error] <0.264.0>@gen_server:error_info/7:934 ** Generic server ejabberd_admin terminating 
** Last message in was {'EXIT',<0.245.0>,shutdown}
** When Server state == {state}
** Reason for termination ==
** {{aborted,{node_not_running,ejabberd@localhost}},
    [{mnesia,abort,1,[{file,"mnesia.erl"},{line,361}]},
     {mnesia_tm,prepare_items,5,[{file,"mnesia_tm.erl"},{line,1225}]},
     {mnesia_tm,dirty,2,[{file,"mnesia_tm.erl"},{line,1067}]},
     {lists,foreach,2,[{file,"lists.erl"},{line,1338}]},
     {ejabberd_commands,unregister_commands,1,
                        [{file,"src/ejabberd_commands.erl"},{line,145}]},
     {gen_server,try_terminate,3,[{file,"gen_server.erl"},{line,718}]},
     {gen_server,terminate,10,[{file,"gen_server.erl"},{line,903}]},
     {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,226}]}]}
2021-02-07 05:38:19.084303+00:00 [error] <0.264.0>@proc_lib:crash_report/4:525 CRASH REPORT:
  crasher:
    initial call: ejabberd_admin:init/1
    pid: <0.264.0>
    registered_name: ejabberd_admin
    exception exit: {aborted,{node_not_running,ejabberd@localhost}}
      in function  mnesia:abort/1 (mnesia.erl, line 361)
      in call from mnesia_tm:prepare_items/5 (mnesia_tm.erl, line 1225)
      in call from mnesia_tm:dirty/2 (mnesia_tm.erl, line 1067)
      in call from lists:foreach/2 (lists.erl, line 1338)
      in call from ejabberd_commands:unregister_commands/1 (src/ejabberd_commands.erl, line 145)
      in call from gen_server:try_terminate/3 (gen_server.erl, line 718)
      in call from gen_server:terminate/10 (gen_server.erl, line 903)
    ancestors: [ejabberd_sup,<0.116.0>]
    message_queue_len: 0
    messages: []
    links: []
    dictionary: []
    trap_exit: true
    status: running
    heap_size: 6772
    stack_size: 28
    reductions: 56958
  neighbours:
2021-02-07 05:38:19.085426+00:00 [error] <0.245.0>@supervisor:do_terminate/2:844 SUPERVISOR REPORT:
    supervisor: {local,ejabberd_sup}
    errorContext: shutdown_error
    reason: {aborted,{node_not_running,ejabberd@localhost}}
    offender: [{pid,<0.264.0>},
               {id,ejabberd_admin},
               {mfargs,{ejabberd_admin,start_link,[]}},
               {restart_type,permanent},
               {shutdown,60000},
               {child_type,worker}]
2021-02-07 05:38:21.166771+00:00 [error] <0.87.0>@supervisor:do_terminate/2:844 SUPERVISOR REPORT:
    supervisor: {local,sasl_safe_sup}
    errorContext: shutdown_error
    reason: killed
    offender: [{pid,<0.88.0>},
               {id,alarm_handler},
               {mfargs,{alarm_handler,start_link,[]}},
               {restart_type,permanent},
               {shutdown,2000},
               {child_type,worker}]

## Bug description

We've observed a crash for our ejabberd server. CPU 2 core RAM 8GB. 
Attaching the memory, CPU, Memory page faults graphs.
cat /proc/sys/fs/file-max 2097152
<img width="869" alt="Screenshot 2021-02-08 at 6 35 20 PM" src="https://user-images.githubusercontent.com/45037304/107223749-8eb8f480-6a3c-11eb-804e-91273d2a33d5.png">
<img width="867" alt="Screenshot 2021-02-08 at 6 35 41 PM" src="https://user-images.githubusercontent.com/45037304/107223760-91b3e500-6a3c-11eb-8c2d-b768b0956406.png">
<img width="861" alt="Screenshot 2021-02-08 at 6 35 47 PM" src="https://user-images.githubusercontent.com/45037304/107223765-924c7b80-6a3c-11eb-9022-1d091d8982c0.png">
<img width="888" alt="Screenshot 2021-02-08 at 6 35 56 PM" src="https://user-images.githubusercontent.com/45037304/107223766-92e51200-6a3c-11eb-8886-7fd9494a81df.png">
<img width="533" alt="Screenshot 2021-02-08 at 6 39 20 PM" src="https://user-images.githubusercontent.com/45037304/107224072-f7a06c80-6a3c-11eb-93ea-a056e1882983.png">

We have enabled muc room persistence but we've not be able to understand if we're missing something to keep memory in control. Looks like GC kicks in only at 75% levels and that sometimes crashes our erlang process.
Any help/pointers to debug further is appreaciated.

**Comment by jnmayur** [comment]
@badlop can you help?

**Comment by Neustradamus** [comment]
@jnmayur: Have you same with ejabberd 21.04?

**Comment by mremond** [comment]
You have a lot of room and I do not think you have enough RAM for that.
I do not think this is a leak. The graph seems to show the standard behaviour of the garbage collector. It triggers when there is more memory pressure.

**Comment by jnmayur** [comment]
@Neustradamus We're on 20.01 & have no plans to upgrade so far.

@mremond If the behaviour is standard and GC kicks in when there is memory pressure, ejabberd shouldn't crash. That was the primary issue we had.

Mentioning a few things we had changed from the native version
1. Changed config or muc rooms to preload_rooms: false. This was done when we realised that by default all rooms are stored in RAM.
2. Changed number of erlang processes ERL_PROCESSES=5000000.

Lately, we have seen issues of GC not kicking in even after consuming 95% RAM and also observed a spike in Memory Page Faults. Refer screenshots.

What in your view would be a good way to estimate RAM and other resource requirements?

<img width="877" alt="Screenshot 2021-04-15 at 3 46 45 PM" src="https://user-images.githubusercontent.com/45037304/114855987-28699d00-9e04-11eb-94e6-3935fcac77d2.png">
<img width="1052" alt="Screenshot 2021-04-15 at 3 49 38 PM" src="https://user-images.githubusercontent.com/45037304/114855998-2c95ba80-9e04-11eb-8436-fd47afb39a6d.png">
